[MERGE] mail, test_mail: remove aggregation from `message_format`

seb-odoo opened the pull request:
Remove aggregation from `message_format `and `_message_read_dict_postprocess`.
There should be no functional change in this PR.

Purpose
=======

The existing code working with "tree" for aggregation and doing explicit "read"
or "search" for related fields does not seem necessary since the ORM of v13.

Considering that it adds complexity to the method, if its original purpose is
not necessary anymore, it would be better to change it.

The notable changes of the ORM that would justify this task are:
 - single cache, shared between current user and sudo (which is called a lot in
   those methods and might explain why the "tree" mechanism was put in place
   originally)
 - m2o and o2m being kept consistent with each other, which allows accessing o2m
   through fields directly (instead of having to use "search" to ensure getting
   consistent data)

Explanation
===========

Doing "read" or "search" could be counter-productive if the data are already in
cache because using those methods will lead to a query no matter what.
The ideal way is to simply access the fields through the records.

As for the "tree" and aggregation in general, the prefetch is automatically
taking care of that when iterating.

task-2180311

alexkuhn commented:
@tde-banana-odoo 
Another PR (https://github.com/odoo/odoo/pull/44170) requires this.
Do you have an estimate on when you will have time for a review? Thanks!

tde-banana-odoo commented:
It is in my pipe. I will try not blocking you too long.

seb-odoo commented on addons/test_mail/tests/test_performance.py:
```diff
@@ -697,6 +697,116 @@ def test_complex_tracking_template(self):
         self.assertEqual(len(rec1.message_ids), 3)
 
+    @mute_logger('odoo.tests', 'odoo.addons.mail.models.mail_mail', 'odoo.models.unlink')
+    @users('__system__', 'emp')
```
What is the deciding factor?

Xavier-Do replied:
We are talking of testing for emp only for a while, usually emp has more query than admin so more coverage, and this method will usually be called by a user I think, not the system. The main goal is to avoid to add useless processing/test time.

seb-odoo replied:
Ok, either way is fine for me, I'll keep just `emp` then

seb-odoo commented on addons/mail/models/mail_message.py:
```diff
@@ -1117,34 +1098,18 @@ def message_format(self):
-                notif_dict[mid]['needaction_partner_ids'].append(notif.res_partner_id.id)
-
         for message in message_values:
+            message_sudo = message_tree[message['id']]
```
Thanks for checking. I tried to go that far but it was more complex to handle, also due to #36263 it will increase the number of queries (o2m are currently more efficient to fetch "in the same _read" as other fields, which is only possibly by using read).

I would be ok to change that too at some point, but I just need the current PR for going forward with mail owl, I can check this again after.

seb-odoo replied:
The related PR has been merged just now, so I will try again now to remove all the remaining `read`.

seb-odoo commented:
Rebased, updated query count, and added a new commit to remove the last remaining `read` and `message_tree`.

Hopefully runbot will be ok with it.

Let me know what you think.

tde-banana-odoo commented on addons/mail/models/mail_message.py:
```diff
@@ -941,92 +941,68 @@ def _notify_moderators(self):
             customer_email_data = []
-            for notification in email_notification_tree[message.id]:
-                customer_email_data.append((partner_tree[notification.res_partner_id.id][0], partner_tree[notification.res_partner_id.id][1], notification.notification_status))
+            filtered_notifications = message_sudo.notification_ids.filtered(lambda n:
```
Couldn't we include a hook in order to fetch other failed notifications ? Currently in SMS we do it in an override of message_format if I am not wrong. While rewriting this one it could be interesting to include it a bit more cleanly. What do you think ?

seb-odoo replied:
I agree, reworking notification is part of the PR that depends on this one, and it is the original reason of doing all of this!

I'll merge it separately but this is almost ready in task-2176017

tde-banana-odoo commented on addons/mail/models/mail_message.py:
```diff
@@ -941,92 +941,68 @@ def _notify_moderators(self):
+            message_sudo = self.browse(vals['id']).sudo().with_prefetch(self.ids)
+
+            # Author
+            if message_sudo.author_id:
```
Couldn't we keep a separation between author and email_from ? That way we don't have fake authors anymore but instead stay on the same logic...

seb-odoo replied:
I'm not sure about the implications of this, and I would prefer not changing it as part of this PR as it would change the result of the method, which I would consider out of scope of this "internal" clean up.

I'll keep it in mind, maybe to include as part of the OWL refactoring.

tde-banana-odoo commented on addons/mail/models/mail_message.py:
```diff
@@ -941,92 +941,68 @@ def _notify_moderators(self):
+            # Attachments
+            main_attachment = self.env['ir.attachment']
+            if message_sudo.attachment_ids and message_sudo.res_id and issubclass(self.pool[message_sudo.model], self.pool['mail.thread']) and has_access_to_model:
+                main_attachment = self.env[message_sudo.model].browse(message_sudo.res_id).message_main_attachment_id
```
What is the purpose of has_access_to_model if you are reading message_main_attachment_id afterwards ? Shoudln't we try / except it instead ?

seb-odoo replied:
This was the old code, I didn't change the logic here. I'll double check if it is still necessary, and either add a comment with my finding or remove it.

seb-odoo replied:
I'm going to remove this condition and do it as sudo:
- from portal everything is sudo so it's pointless to check access on top of it
- from backend, it's almost impossible to trigger the case, and it's not like returning is_main True would leak any sensitive information to an employee

tde-banana-odoo commented on addons/test_mail/tests/test_performance.py:
```diff
@@ -696,6 +696,119 @@ def test_complex_tracking_template(self):
+        name_field = self.env['ir.model.fields']._get(self.umbrella._name, 'name')
+        customer_id_field = self.env['ir.model.fields']._get(self.umbrella._name, 'customer_id')
+
+        messages = self.env['mail.message'].sudo().create([{
```
Isn't it possible to shorten a bit data creation ? 80 lines for 2 messages, looks like JS tests. Some data is redundant, could be done using lists, ...

seb-odoo replied:
I don't really see how to reduce it while keeping the current coverage, but if you have the time to give me hints I'm open to shorten it.

Xavier-Do replied:
Example: 
```
            'notification_ids': [
                (0, 0, {
                    'res_partner_id': self.partners[3].id,
                    'notification_type': 'inbox',
                }),
                (0, 0, {
                    'res_partner_id': self.partners[4].id,
                    'notification_type': 'email',
                    'notification_status': 'exception',
                }),
                (0, 0, {
                    'res_partner_id': self.partners[6].id,
                    'notification_type': 'email',
                    'notification_status': 'exception',
                }),
            ],
```

```
'notification_ids': [ (0, 0, {
                    'res_partner_id': self.partners[partner_index].id,
                    'notification_type': n_type,
                    'notification_status': n_status,
                }) for (partner_index, n_type, n_status) in [(3, 'inbox', 'ready'), (4, 'email', 'exception'),  (5, 'email, exception)', ]]
```
(Not perfect, but an example of a way to shorten it)

tde-banana-odoo commented on odoo/models.py:
```diff
@@ -2932,6 +2932,9 @@ def read(self, fields=None, load='_classic_read'):
 
+        return self._read_format(fields=fields, load=load)
+
+    def _read_format(self, fields=None, load='_classic_read'):
```
I don't understand "read_format" naming. How is any formatting linked with the method itself ? As it seems that _read populates the cache, maybe we could imagine something like _read_from_cache ? Just proposing :) .

seb-odoo replied:
The main reason of extracting the method is that is does `convert`, which calls `convert_to_read` on every field.

This is important in our case because for relational fields we want to return a pretty list of ids and names, not a recordset. Some other fields also have specific formatting.

So in a way it is "formatting" the result of a `read` for a specific purpose (sending to JS).

Calling it `_read_from_cache` would be ok for me too, considering we implicitly think of `read` as already "formatting" the result too.

Note: I changed the signature of the method since your comment because not giving `fields` is not supported here, it can't default to None.

rco-odoo commented on addons/mail/models/mail_message.py:
```diff
@@ -941,92 +941,68 @@ def _notify_moderators(self):
-    def _message_read_dict_postprocess(self, messages, message_tree):
-        """ Post-processing on values given by message_read. This method will
-            handle partners in batch to avoid doing numerous queries.
+    def _message_read_dict_postprocess(self, vals_list):
```
Why don't you pass the list of fields, and read the data on `self`?
You can then format the fields by using `field.convert_to_read()` or any formatting you like.

rco-odoo replied:
It would be similar to calling `_read_format` *after* `_message_read_dict_postprocess` instead of *before* it.

seb-odoo replied:
As discussed, ok to move `_read_format` inside the method, but it still has to be called at the start and not at the end since it reads fields specifically and then we apply changes on top of that (of keys that are fields but also on new keys that are not fields).

rco-odoo commented on addons/mail/models/mail_message.py:
```diff
@@ -941,92 +941,68 @@ def _notify_moderators(self):
+
+            # Author
+            if message_sudo.author_id:
+                author = message_sudo.author_id.name_get()[0]
```
```suggestion
                author = message_sudo.author_id.display_name
```

seb-odoo replied:
ok

rco-odoo commented on addons/mail/models/mail_message.py:
```diff
@@ -941,92 +941,68 @@ def _notify_moderators(self):
+                (n.notification_status in ('bounce', 'exception', 'canceled') or n.res_partner_id.partner_share)
+            )
+            for notification in filtered_notifications:
+                partner_name_get = notification.res_partner_id.name_get()[0]
```
You could use `display_name` instead of `name_get`.

seb-odoo replied:
ok

seb-odoo commented:
Thank you for the reviews, changes have been applied accordingly.

seb-odoo commented:
@robodoo r+

seb-odoo commented:
@robodoo rebase-merge

robodoo commented:
Merge method set to rebase and merge, using the PR as merge commit message

Xavier-Do commented on addons/mail/models/mail_message.py:
```diff
@@ -941,100 +941,71 @@ def _notify_moderators(self):
+        """Reads values from messages and formats them for the web client."""
+        self.check_access_rule('read')
+        vals_list = self._read_format(fnames)
         safari = request and request.httprequest.user_agent.browser == 'safari'
```
move this line where we need it.

seb-odoo replied:
Exactly, I've had times where accessing data on `request` was slower than it looks due to magic property, so I prefer to keep it outside the loop.

Xavier-Do commented on addons/test_mail/tests/test_performance.py:
```diff
@@ -267,17 +267,17 @@ def test_mail_composer(self):
             })
 
-        with self.assertQueryCount(__system__=36, emp=41):
+        with self.assertQueryCount(__system__=31, emp=37):
```
a multibuild would be great to validate query count changes

seb-odoo replied:
Hopefully the only remaining issue is the query that is marked as non deterministic, for which I have been extra careful.

If there are others discovered since this merge feel free to ping me and I'll handle them.

Xavier-Do commented on addons/test_mail/tests/test_performance.py:
```diff
@@ -267,17 +267,17 @@ def test_mail_composer(self):
         test_template = self.env['mail.template'].browse(self.test_template_full.id)
-        # TODO XDO/TDE/MAT FIXME non deterministic between 25 and 29 queries
-        with self.assertQueryCount(__system__=28, emp=29):
+        # TODO FIXME non deterministic, last check 25 feb 2020. Runbot at most +7 compared to local.
```
I need to take some time to work on that

seb-odoo replied:
Let me know if I can help, I'll be glad when this is resolved.

Xavier-Do commented on addons/test_mail/tests/test_performance.py:
```diff
@@ -253,7 +253,7 @@ def test_adv_activity_mixin(self):
     @users('__system__', 'emp')
     @warmup
-    @mute_logger('odoo.addons.mail.models.mail_mail')
+    @mute_logger('odoo.tests', 'odoo.addons.mail.models.mail_mail')
```
why?

seb-odoo replied:
As seen in first commit, this is to have a cleaner log, similar to what is done for other tests in the field (eg. `test_complex_mail_mail_send`).

This particular case is to avoid `odoo.tests: skip sending email in test mode`

Xavier-Do replied:
Ho, great!
